cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherServerHandler.java

```diff
@@ -0,0 +1,273 @@
+    throws NotImplementedException, IOException {
+    checkTetherServerEnabled();
+
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetherConnectionRequest tetherRequest = GSON.fromJson(content, TetherConnectionRequest.class);
+    TopicId topicId = new TopicId(NamespaceId.SYSTEM.getNamespace(),
+                                  TETHERING_TOPIC_PREFIX + tetherRequest.getPeer());
+    try {
+      messagingService.createTopic(new TopicMetadata(topicId, Collections.emptyMap()));
+    } catch (TopicAlreadyExistsException e) {
+      LOG.warn("Topic {} already exists", topicId);
+    } catch (IOException e) {
+      LOG.error("Failed to create topic {}", topicId, e);
+      throw e;
+    }
+
+    try {
+      store.getPeer(tetherRequest.getPeer());
+      // Peer is already configured, treat this as a no-op.
+      responder.sendStatus(HttpResponseStatus.OK);
+      return;
+    } catch (PeerNotFoundException ignored) {
+    }
+
+    // We don't need to keep track of the client metadata on the server side.
+    PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaceAllocations(), Collections.emptyMap());
+    // We don't store the peer endpoint on the server side because the connection is initiated by the client.
+    PeerInfo peerInfo = new PeerInfo(tetherRequest.getPeer(), null, TetherStatus.PENDING, peerMetadata);
+    try {
+      store.addPeer(peerInfo);
+    } catch (Exception e) {
+      try {
+        messagingService.deleteTopic(topicId);
+      } catch (Exception ex) {
+        e.addSuppressed(ex);
+      }
+      throw new IOException("Failed to create tether with peer " + tetherRequest.getPeer(), e);
+    }
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Accepts/rejects the tethering request.
+   */
+  @POST
+  @Path("/tethering/connections/{peer}")
+  public void tetherAction(HttpRequest request, HttpResponder responder, @PathParam("peer") String peer,
+                           @QueryParam("action") String action)
+    throws NotImplementedException, BadRequestException, PeerNotFoundException, IOException {
+    TetherStatus tetherStatus;
+    switch (action) {
+      case "accept":
+        tetherStatus = TetherStatus.ACCEPTED;
+        break;
+      case "reject":
+        tetherStatus = TetherStatus.REJECTED;
+        break;
+      default:
+        throw new BadRequestException(String.format("Invalid action: %s", action));
+    }
+    updateTetherStatus(responder, peer, tetherStatus);
+  }
+
+  private void initializeMessageIds() {
+    lastMessageIds = new HashMap<>();
+    List<String> topics;
+    try {
+      topics = store.getPeers().stream()
+        .filter(p -> p.getTetherStatus() == TetherStatus.ACCEPTED)
+        .map(PeerInfo::getName)
+        .collect(Collectors.toList());
+    } catch (IOException e) {
+      LOG.warn("Failed to get peer information", e);
+      return;
+    }
+
+    for (String topic: topics) {
```

Comment:
This loop creates many transactions. It is better to do all the lookup in the same transaction.

Reply:
this code was removed in the latest update